### ISSUE TITLE
Why does test runner hang when browser location is manipulated?

### ISSUE THREAD

**Issue by leonsenft** [dart-lang/test#928]
Consider the following test:

```dart
@TestOn('browser')
import 'dart:html';

import 'package:test/test.dart';

void main() {
  test('pushState()', () {
    // Update the current browser location.
    window.history.pushState(null, null, '/foo');
  });
}
```

If I run this test on the command-line using

```
pub run test -p chrome
```

or

```
pub run build_runner test -- -p chrome
```

the test case hangs indefinitely. However, if I serve the test

```
pub run build_runner serve test
```

and click the test link in the launched browser session, the path is updated to `/foo` and the test passes with the following console output:

```
Navigated to http://localhost:8080/browser_history_test.debug.html
dart_sdk.js:18257 00:00 +0: pushState()
dart_sdk.js:18257 00:00 +1: All tests passed!
```

Does manipulating the browser location affect the command-line test runner's ability to communicate with the running test and receive the results? The test runner uses a web socket connection to communicate with the browser tests, which I believe should persist through location changes triggered by `pushState()` or `replaceState()`.

Or does the test runner intentionally ignore test results that don't originate from the location at which the test began?

**Comment by natebosch** [comment]
The test runner runs (multiple concurrent?) tests in iframes and so it's not tolerant to location changes. I'm not sure how easy it would be to change that at this point...

**Comment by matanlurey** [comment]
I guess theoretically it could be "fixed" with `-j=1`? This definitely makes it (extra) painful to do any sort of extensive testing of SPA-type web applications without always mocking-out the history object (in this case @leonsenft is actually writing tests for the _real_ history object).

**Comment by grouma** [comment]
I did some digging. I believe it is related to this issue: https://github.com/dart-lang/sdk/issues/22554

We seem to drop messages on the floor if they don't have the same src:
https://github.com/dart-lang/test/blob/master/tool/host.dart#L210

To be honest I'm not sure why that is necessary. The internal communication is much simpler as we don't have to worry about multiple iFrames so I haven't dealt with host.dart much.

**Comment by grouma** [comment]
@nex3 Natalie can you shed some light?

**Comment by leonsenft** [comment]
@natebosch Thanks for the context, that does make sense.

@matanlurey Just tried with `-j 1` and it still hangs, although maybe that argument could *allow* the runner to lift this restriction if it were programmed that to do so?

@grouma I suspect that's the cause, thanks for investigating.

Like Matan said, in any actual SPA test, you would fake your browser history manipulations to avoid these kind of issues.

My use case is rather particular, since I'm *writing* a package for history manipulation, which includes the aforementioned fake implementation for testing. Unlike the users of the this package, I actually need to test the real implementations. I was hoping I could get away with writing purely in-browser tests, but this limitation means I'll probably have to write web-driver tests.

**Comment by nex3** [comment]
@grouma With the caveat that I don't have all the test context loaded into my head anymore, I believe you're right: we check the href of messages from iframes as a security measure to guard against other webapps sending messages to the test runner and using it to snoop on your local filesystem. This could be solved by using a `MessageChannel` to communicate between the iframe and the main frame, since this guarantees the source of the messages, but at the time that code was written doing so was blocked on https://github.com/dart-lang/sdk/issues/26349.

**Comment by leonsenft** [comment]
Thanks for the response @nex3.

I took a shot at fixing this using a proper `MessageChannel` and hit another issue with `dart:html`: [`MessageEvent`](https://api.dartlang.org/stable/2.0.0/dart-html/MessageEvent-class.html) does not expose the [`ports`](https://developer.mozilla.org/en-US/docs/Web/API/MessageEvent/ports) property.

For context as to why this matters, in [`_connectToIframe()`](https://github.com/dart-lang/test/blob/master/tool/host.dart#L201) we'd like to initialize a [`MessageChannel`](https://api.dartlang.org/stable/2.0.0/dart-html/MessageChannel-class.html) to use for communicating between the host script and the iframe.

The steps to establish this communication...

...from host:
1. Create `MessageChannel`
2. Listen for messages on `MessageChannel.port1`
3. Send `MessageChannel.port2` to iframe via `iframe.contentWindow.postMessage()`

...from iframe:
1. Listen for first `MessageEvent` on `window.onMessage`
2. Get `MessagePort` from `MessageEvent.ports` (this is the `MessageChannel.port2` we sent via `iframe.contentWindow.postMessage()`)
3. Listen for messages on this `MessagePort`

Since `dart:html` doesn't expose `MessageEvent.ports`, we can't access the `MessagePort` from the iframe.

**Comment by leonsenft** [comment]
Blocked by https://github.com/dart-lang/sdk/issues/32688.

**Comment by leonsenft** [comment]
Duplicate of https://github.com/dart-lang/test/issues/562.
